File: app/src/main/java/ru/omsu/themoviedb/Adapters/MovieAdapter.java
package ru.omsu.themoviedb.Adapters;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import ru.omsu.themoviedb.Metadata.TMDB.Item.Movie.Movie;
import ru.omsu.themoviedb.Metadata.TMDB.Item.Movie.MoviesPage;
import ru.omsu.themoviedb.Metadata.TMDB.TMDBService;
import ru.omsu.themoviedb.R;
import ru.omsu.themoviedb.UI.Activities.ItemInfoActivity;

import static ru.omsu.themoviedb.Settings.API_KEY;
import static ru.omsu.themoviedb.Settings.POPULAR;
import static ru.omsu.themoviedb.Settings.TOP_RATED;
import static ru.omsu.themoviedb.Settings.UPCOMING;
import static ru.omsu.themoviedb.Settings.URL_TMDB_COMPRESSED_IMAGE;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieViewHolder> {

    private int page = 1;
    private int category = POPULAR;
    private Context context;

    private List<Movie> movieList = Collections.synchronizedList(new ArrayList<Movie>());


    public MovieAdapter(Context context) {
        this.context = context;
    }

    public int getCategory() {
        return category;
    }

    private void setCategory(int category) {
        this.category = category;
    }

    @Override
    public void onBindViewHolder(MovieViewHolder holder, int position) {
        if ((movieList.size() - position) <= 10) {
            getMovies();
        }
        holder.bind(movieList.get(position));
    }

    @Override
    public MovieViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.card_item, parent, false);
        return new MovieViewHolder(view);
    }

    @Override
    public int getItemCount() {
        return movieList.size();
    }

    private void resetPage(){
        this.page = 1;
    }

    public void clearAndSetPopular(){
        clearItems();
        setCategory(POPULAR);
        resetPage();
        getMovies();

    }
    
    public void clearAndSetUpcoming(){
        clearItems();
        setCategory(UPCOMING);
        resetPage();
        getMovies();
    }
    
    public void clearAndSetTopRated(){
        clearItems();
        setCategory(TOP_RATED);
        resetPage();
        getMovies();
    }

    public void setItems(Collection<Movie> movies) {
        movieList.addAll(movies);
        notifyDataSetChanged();
    }

    private void clearItems() {
        movieList.clear();
        notifyDataSetChanged();
    }

    public void getMovies(){
        switch (category) {
            case POPULAR:
                TMDBService
                        .getInstance()
                        .getJSONApi()
                        .getPopularMovies(API_KEY, Locale.getDefault().toString().replace("_", "-"), page, Locale.getDefault().getCountry())
                        .enqueue(new Callback<MoviesPage>() {
                            @Override
                            public void onResponse(Call<MoviesPage> call, retrofit2.Response<MoviesPage> response) {
                                MoviesPage moviesPage = response.body();
                                page++;
                                movieList.addAll(moviesPage.results);
                                notifyDataSetChanged();
                            }

                            @Override
                            public void onFailure(Call<MoviesPage> call, Throwable t) {
                            }
                        });
                break;
            case TOP_RATED:
                TMDBService
                        .getInstance()
                        .getJSONApi()
                        .getTopRatedMovies(API_KEY, Locale.getDefault().toString().replace("_", "-"), page, Locale.getDefault().getCountry())
                        .enqueue(new Callback<MoviesPage>() {
                            @Override
                            public void onResponse(Call<MoviesPage> call, retrofit2.Response<MoviesPage> response) {
                                MoviesPage moviesPage = response.body();
                                page++;
                                movieList.addAll(moviesPage.results);
                                notifyDataSetChanged();
                            }
                            @Override
                            public void onFailure(Call<MoviesPage> call, Throwable t) {
                            }
                        });
                break;
            case UPCOMING:
                TMDBService
                        .getInstance()
                        .getJSONApi()
                        .getUpcomingMovies(API_KEY, Locale.getDefault().toString().replace("_", "-"), page, Locale.getDefault().getCountry())
                        .enqueue(new Callback<MoviesPage>() {
                            @Override
                            public void onResponse(Call<MoviesPage> call, retrofit2.Response<MoviesPage> response) {
                                MoviesPage moviesPage = response.body();
                                page++;
                                movieList.addAll(moviesPage.results);
                                notifyDataSetChanged();
                            }

                            @Override
                            public void onFailure(Call<MoviesPage> call, Throwable t) {
                            }
                        });
                break;
        }


    }

    class MovieViewHolder extends RecyclerView.ViewHolder {
        private ImageView posterView;
        private ImageView backgroundView;
        private TextView titleView;
        private TextView releaseDateView;
        private TextView overviewView;
        private TextView ratingView;

        MovieViewHolder(View itemView) {
            super(itemView);
            posterView = itemView.findViewById(R.id.poster);
            backgroundView = itemView.findViewById(R.id.background);
            titleView = itemView.findViewById(R.id.title);
            releaseDateView = itemView.findViewById(R.id.releaseDate);
            overviewView = itemView.findViewById(R.id.overview);
            ratingView = itemView.findViewById(R.id.rating);
        }

        void bind(final Movie movie) {
            titleView.setText(movie.getTitle());
            overviewView.setText(movie.getOverview());
            if (movie.getVoteAverage() != 0)
                ratingView.setText(movie.getVoteAverage().toString());
            else ratingView.setText("-");
            releaseDateView.setText(movie.getReleaseDate());
            if (!Objects.equals(movie.getPosterPath(), null))
                Glide.with(itemView.getContext()).load(URL_TMDB_COMPRESSED_IMAGE + movie.getPosterPath()).fitCenter().dontTransform().into(posterView);
            if (!Objects.equals(movie.getBackdropPath(), null))
                Glide.with(itemView.getContext()).load(URL_TMDB_COMPRESSED_IMAGE + movie.getBackdropPath()).fitCenter().dontTransform().into(backgroundView);
            posterView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(context, ItemInfoActivity.class);
                    intent.putExtra("MOVIE_ID", movie.getId());
                    context.startActivity(intent);
                }
            });
            backgroundView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(context, ItemInfoActivity.class);
                    intent.putExtra("MOVIE_ID", movie.getId());
                    context.startActivity(intent);
                }
            });
        }
    }
}

File: app/src/main/java/ru/omsu/themoviedb/adapters/MovieAdapter.kt
package ru.omsu.themoviedb.adapters

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import ru.omsu.themoviedb.R
import ru.omsu.themoviedb.Settings.API_KEY
import ru.omsu.themoviedb.Settings.BACKDROP_SIZE_W780
import ru.omsu.themoviedb.Settings.POSTER_SIZE_W500
import ru.omsu.themoviedb.Settings.URL_TMDB_BASE
import ru.omsu.themoviedb.UI.Activities.ItemInfoActivity
import ru.omsu.themoviedb.enums.ItemType
import ru.omsu.themoviedb.enums.RequestTypeMovies
import ru.omsu.themoviedb.metadata.tmdb.TMDBService
import ru.omsu.themoviedb.metadata.tmdb.movie.Movie
import java.util.*

class MovieAdapter(private val context: Context) : RecyclerView.Adapter<MovieAdapter.MovieViewHolder>() {

    private var page = 1
    var category = RequestTypeMovies.POPULAR
        private set

    private val movieList = Collections.synchronizedList(ArrayList<Movie>())

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        if (movieList.size - position <= 10) {
            getMovies()
            page++
        }
        holder.bind(movieList[position])
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val view = LayoutInflater
                .from(parent.context)
                .inflate(R.layout.card_item, parent, false)
        return MovieViewHolder(view)
    }

    override fun getItemCount(): Int {
        return movieList.size
    }

    private fun resetPage() {
        this.page = 1
    }

    fun clearAndSetPopular() {
        clearItems()
        category = RequestTypeMovies.POPULAR
        resetPage()
        getMovies()

    }

    fun clearAndSetUpcoming() {
        clearItems()
        category = RequestTypeMovies.UPCOMING
        resetPage()
        getMovies()
    }

    fun clearAndSetTopRated() {
        clearItems()
        category = RequestTypeMovies.TOP_RATED
        resetPage()
        getMovies()
    }

    fun setItems(movie: Collection<Movie>) {
        movieList.addAll(movie)
        notifyDataSetChanged()
    }

    private fun clearItems() {
        movieList.clear()
        notifyDataSetChanged()
    }

    fun getMovies() {
        val tmdbService = TMDBService.create()
        tmdbService.getPopularMovies(API_KEY, Locale.getDefault().toString().replace("_", "-"), page, "RU")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ result ->
                    movieList.addAll(result.results!!)
                    notifyDataSetChanged()
                }, { error ->
                    error.printStackTrace()
                })
    }


    inner class MovieViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val posterView: ImageView = itemView.findViewById(R.id.poster)
        private val backgroundView: ImageView = itemView.findViewById(R.id.background)
        private val titleView: TextView = itemView.findViewById(R.id.title)
        private val releaseDateView: TextView = itemView.findViewById(R.id.releaseDate)
        private val overviewView: TextView = itemView.findViewById(R.id.overview)
        private val ratingView: TextView = itemView.findViewById(R.id.rating)

        fun bind(movie: Movie) {
            titleView.text = movie.title
            overviewView.text = movie.overview
            if (movie.vote_average != 0.0) {
                ratingView.text = movie.vote_average.toString()
            } else
                ratingView.text = "-"
            releaseDateView.text = movie.release_date
            Glide.with(itemView.context).load(URL_TMDB_BASE + POSTER_SIZE_W500 + movie.poster_path).fitCenter().dontTransform().into(posterView)
            Glide.with(itemView.context).load(URL_TMDB_BASE + BACKDROP_SIZE_W780 + movie.backdrop_path).fitCenter().dontTransform().into(backgroundView)
            posterView.setOnClickListener {
                val intent = Intent(context, ItemInfoActivity::class.java)
                intent.putExtra("ITEM_TYPE", ItemType.MOVIE)
                intent.putExtra("ITEM_ID", movie.id)
                context.startActivity(intent)
            }
            backgroundView.setOnClickListener {
                val intent = Intent(context, ItemInfoActivity::class.java)
                intent.putExtra("ITEM_TYPE", ItemType.MOVIE)
                intent.putExtra("ITEM_ID", movie.id)
                context.startActivity(intent)
            }
        }
    }
}
